### ISSUE TITLE
If building for arm fails, will build-bootstrap.sh rebuild the already-built aarch64 pkgs from scratch?

### ISSUE THREAD

**Issue by segfault-bilibili** [termux/termux-packages#12529]
### Discussed in https://github.com/termux/termux-packages/discussions/12525

<div type='discussions-op-text'>

<sup>Originally posted by **segfault-bilibili** October 24, 2022</sup>
I see `arm` pkgs are built after `aarch64` ones by default. Yesterday I encountered an error during building `arm` pkgs, then it seemed that `build-bootstrap.sh` started to rebuild all (already-built) `aarch64` pkgs from scratch as well, which does not seem favorable.</div>

**Comment by segfault-bilibili** [comment]
To me this is unexpected. When seeing an error, my natural reaction is to restart the building process, with an expectation that the building process would just continue from where things didn't succeed. Maybe I should not have that expectation at all?

In my opinion, maybe there should be something like a marker which records which architectures have been successfully built so that they won't have to be rebuilt from scratch.

**Comment by segfault-bilibili** [comment]
I also had another idea that the building directory should contain the architecture, so that it will (probably) naturally mark itself as "successfully built", without being overwritten by a different architecture. However this idea seems to hog much more disk space.

**Comment by segfault-bilibili** [comment]
Sorry, I didn't notice `TERMUX_BUILT_PACKAGES_DIRECTORY_FOR_ARCH`, which seems to be what I want - however it didn't seem to work as intended for me.

**Comment by segfault-bilibili** [comment]
Oh sorry again, `TERMUX_BUILT_PACKAGES_DIRECTORY_FOR_ARCH` is probably different from what I thought of. I thought that the entire architecture (instead of individual packages) could be skipped after being successfuly built.

**Comment by Grimler91** [comment]
Maybe you are looking for the `--architectures` flag? Use `./script/build-bootstrap.sh --architectures arm` to (re)build just for arm

**Comment by agnostic-apollo** [comment]
Packages would not get rebuilt for architecture by default unless `-f` flag is passed. But in the current broken state of  `build-bootstrap.sh`, the  `-f` flag is necessary to build a correct bootstrap zip.

Check https://github.com/termux/termux-packages/pull/10296#issuecomment-1108052222 and https://github.com/termux/termux-packages/pull/10540#issuecomment-1264497128

**Comment by agnostic-apollo** [comment]
>is it possible to generate bootstrap.zip generate-bootstrap.sh, with the locally built debs, instead of downloading them online? 

The `build-bootstrap.sh` will not download any debs by default. Everything is built locally, even if building for `com.termux` since `build-packge.sh` is not passed the `-i`/`-I` flags.

> Do I need to re-run build-bootstrap.sh with -f flag with each individual architecture 

Only `-f` is needed, whether you pass one or all architectures. Single run should work. But if failure occurs, all already build packages/architectures will need to be built again due to `-f` flag.

>seems to include some seemingly unrelated packages

Yeah, that's another issue I wrote about at https://github.com/termux/termux-app/issues/2892#issuecomment-1186482737

You can call `extract_debs` and `create_bootstrap_archive` manually. For now the additional package issue can be temporarily be solved by adding a `--skip-packages` option which is used by `build-package.sh`, will try taking out time for it today. I also currently don't really have free space to build complete archives locally on my laptop, so that's another issue.

https://github.com/termux/termux-packages/blob/c96f4a7942d13068e06b1c1e562f344f70cfbf94/scripts/build-bootstraps.sh#L93

**Comment by segfault-bilibili** [comment]
Oh it seems that all built packages of architectures other than the last one (x86_64) are deleted. I don't know whether it's related to `-f` - a quick CTRL-F doen't find anything related.

**Comment by agnostic-apollo** [comment]
> however it seems that all built packages of architectures other than the last one (x86_64) are deleted. I

When you **start** building debs for an arch, any existing packages and debs for all archs are deleted if `-f` is passed. No deletion happens at the **end** of builds, so built packages of last arch remain.

https://github.com/termux/termux-packages/blob/c96f4a7942d13068e06b1c1e562f344f70cfbf94/scripts/build-bootstraps.sh#L370-L373

**Comment by segfault-bilibili** [comment]
Probably I have got what I want.

I modified `generate-bootstraps.sh` to read `/var/lib/dpkg/status` (of unzipped `boostrap.zip` of each architecture) instead of downloading `Package` file online.

*(I didn't push my modifications though. just pointing out the related position in code)*

https://github.com/termux/termux-packages/blob/7269ff9395b6bf11bec0c40aba456bca75ed8861/scripts/generate-bootstraps.sh#L60-L69

https://github.com/termux/termux-packages/blob/7269ff9395b6bf11bec0c40aba456bca75ed8861/scripts/build-bootstraps.sh#L171-L177

However I encountered two problems:

1. `/var/lib/dpkg/status` lacks the `Filename:` field which is required in `generate-bootstraps.sh` here:
https://github.com/termux/termux-packages/blob/7269ff9395b6bf11bec0c40aba456bca75ed8861/scripts/generate-bootstraps.sh#L123
Fortunately this seems possible to work around: I can still concatenate name, version, and architecture of the package, which seem to be all I need to reconstruct missing the deb filename.

2. (a known issue probably?) `boostrap.zip`s of `arm` and `i686` seem to have incorrect `Architecture:` in their own `/var/lib/dpkg/status`, which were all incorrectly set to `aarch64` instead of their own correct values respectively. I did a manual `sed -i` to fix this.
